File: sdk/gsusb-cpp/examples/01_simple_std.cpp
#include "demo_common.hpp"

#include <iostream>

int main(int argc, char** argv) {
    try {
        const auto options = demo::parse_common_options(argc, argv, "Single-channel classic CAN example");
        std::cout << "Opening selector=" << options.selector << " tx=" << options.tx_channel << " rx=" << options.rx_channel << " @ " << options.bitrate << "bps\n";
        auto tx_bus = std::make_unique<gsusb::GsUsbBus>(demo::make_bus_options(options, options.tx_channel, false));
        auto rx_bus = std::make_unique<gsusb::GsUsbBus>(demo::make_bus_options(options, options.rx_channel, false));

        gsusb::CanMessage message;
        message.arbitration_id = 0x123;
        message.data = {0x11, 0x22, 0x33, 0x44};

        demo::print_message("Sending", message);
        tx_bus->send(message);
        auto received = rx_bus->recv(std::chrono::seconds(2));
        if (!received.has_value()) {
            std::cerr << "No frame received\n";
            return 1;
        }
        demo::print_message("Received", received.value());
        tx_bus->shutdown();
        rx_bus->shutdown();
        return 0;
    } catch (const std::exception& ex) {
        std::cerr << ex.what() << "\n";
        return 1;
    }
}

File: sdk/gsusb-csharp/examples/GsUsb.Examples/Examples/01_simple_std.cs
using GsUsb;

namespace GsUsb.Examples;

internal static class Example01SimpleStd
{
    public static int Run(string[] args)
    {
        var options = CommonOptions.Parse(args, "Single-channel classic CAN example");
        Console.WriteLine($"Opening {options.Selector} tx={options.TxChannel} rx={options.RxChannel} @ {options.Bitrate}bps");
        using var txBus = new GsUsbBus(DemoCommon.MakeBusOptions(options, options.TxChannel, fd: false));
        using var rxBus = new GsUsbBus(DemoCommon.MakeBusOptions(options, options.RxChannel, fd: false));

        var message = new CanMessage
        {
            ArbitrationId = 0x123,
            Data = [0x11, 0x22, 0x33, 0x44],
        };

        DemoCommon.PrintMessage("Sending", message);
        txBus.Send(message);
        var received = rxBus.Receive(TimeSpan.FromSeconds(2));
        if (received is null)
        {
            Console.Error.WriteLine("No frame received");
            return 1;
        }

        DemoCommon.PrintMessage("Received", received);
        return 0;
    }
}